openstf/feature_engineering/lag_features.py

```diff
@@ -0,0 +1,152 @@
+# SPDX-FileCopyrightText: 2017-2021 Alliander N.V. <korte.termijn.prognoses@alliander.com> # noqa E501>
+#
+# SPDX-License-Identifier: MPL-2.0
+
+import re
+
+import numpy as np
+import scipy.signal
+
+
+def generate_lag_feature_functions(feature_set_names=None, horizon=24.0):
+    """Creates functions to generate lag features in a dataset.
+
+    Args:
+        feature_set_names (list of strings): minute lagtimes that where used during training
+            of the model. If empty a new set will be automatically generated.
+        horizon (float): Forecast horizon limit in hours.
+
+
+    Returns:
+        dict: dictionary with lag functions
+
+    Example:
+        lag_functions = generate_lag_functions(data,minute_list,h_ahead)
+    """
+
+    # Generate lag_times if no features are provided
+    if feature_set_names is None:
+        lag_times_minutes, lag_time_days_list = generate_trivial_lag_features(horizon)
+
+    # Or extract lag features if provided
+    else:
+        lag_times_minutes, lag_time_days_list = extract_lag_features(feature_set_names)
+
+    # Empty dict to store all generated lag functions
+    lag_functions = {}
+
+    # Add intraday-lag functions (lags in minutes)
+    for minutes in lag_times_minutes:
+
+        def func(x, shift=minutes):
+            return x.shift(freq="T", periods=1 * shift)
+
+        new = {"T-" + str(int(minutes)) + "min": func}
+        lag_functions.update(new)
+
+    # Add day lag functions:
+    for day in lag_time_days_list:
+
+        def func(x, shift=day):
+            return x.shift(freq="1d", periods=1 * shift)
+
+        new = {"T-" + str(int(day)) + "d": func}
+        lag_functions.update(new)
+    return lag_functions
+
+
+def generate_trivial_lag_features(horizon):
+    """Function that generates relevant lag times for lag feature function creation.
+    This function is mostly used during training of models and not during predicting
+    Args:
+        horizon: Forecast horizon limit in hours.
+
+    Returns:
+        minutes_list (list[int]): list of minute lags that were used as features during training
+        days_list (list[int]): list of minute lags that were used as features during training
+
+    """
+    mindays = int(np.ceil(horizon / 24))
+    lag_time_days_list = list(np.linspace(mindays, 14, 15 - mindays))
+
+    # Make list of trivial lag times
+    trivial_lag_minutes_list = np.linspace(60, 24 * 60, 24).tolist() + [15, 30, 45]
+
+    # Discard lag times that are not available for the specified horizon
+    trivial_lag_times_minutes = set(
+        [i for i in trivial_lag_minutes_list if i > horizon * 60]
+    )
+
+    return trivial_lag_times_minutes, lag_time_days_list
+
+
+def extract_lag_features(lag_features):
+    """Creates a list of lag minutes and a list of lag days that were used during
+    the training of the input model
+
+    Args:
+        lag_features (list[str]): All requested lag features
```

Comment:
@JanMaartenvanDoorn I would not use `lag_features` as the argument here. You don't use the function in that way and it does not matter. In other places you call this thing `feature_set_names` or `feature_set_list`. Now we have three names for the same thing.

Reply:
Thinking about it why not use `feature_names` ? That's what is is I guess?